scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingMethodImpl.java

```diff
@@ -166,50 +167,102 @@ public int schedule() {
 
         Map<JobId, JobDescriptor> toUnlock = jobMap;
 
-        if (logger.isTraceEnabled() && jobMap == null || jobMap.isEmpty()) {
-            logger.trace("No jobs selected to be scheduled");
-        }
-        if (logger.isDebugEnabled() && !jobMap.isEmpty()) {
-            logger.debug("jobs selected to be scheduled : " + jobMap);
-        }
+        logSelectedJobs(jobMap);
 
         // If there are some jobs which could not be locked it is not possible to do any priority scheduling decision,
-        // we wait for next scheduling loop
+        // we wait for next scheduling loop and don't start any task
         if (jobMap.isEmpty()) {
-            return numberOfTaskStarted;
+            return 0;
         }
 
+        return startTasks(currentPolicy, jobMap, toUnlock);
+    }
+
+    private int startTasks(Policy currentPolicy, Map<JobId, JobDescriptor> jobMap, Map<JobId, JobDescriptor> toUnlock) {
         try {
             List<JobDescriptor> descriptors = new ArrayList<>(jobMap.values());
 
             //get rmState and update it in scheduling policy
-            RMState rmState = getRMProxiesManager().getRmProxy().getState();
-            currentPolicy.setRMState(rmState);
-            Set<String> freeResources = rmState.getFreeNodes();
-            if (logger.isDebugEnabled()) {
-                logger.debug("eligible nodes : " + freeResources);
-            }
-            //if there is no free resources, stop it right now
+            Set<String> freeResources = getFreeResources(currentPolicy);
+            //if there is no free resources, stop it right now without starting any task
             if (freeResources.isEmpty()) {
-                return numberOfTaskStarted;
+                return 0;
             }
 
             // ask the policy all the tasks to be schedule according to the jobs list.
+            LinkedList<EligibleTaskDescriptor> fullListOfTaskRetrievedFromPolicy = currentPolicy.getOrderedTasks(descriptors);
 
-            LinkedList<EligibleTaskDescriptor> taskRetrievedFromPolicy = currentPolicy.getOrderedTasks(descriptors);
+            //if there is no task to scheduled, return without starting any task
+            if (fullListOfTaskRetrievedFromPolicy == null || fullListOfTaskRetrievedFromPolicy.isEmpty()) {
+                return 0;
+            }
+
+            toUnlock = unlockResources(toUnlock);
 
-            //if there is no task to scheduled, return
-            if (taskRetrievedFromPolicy == null || taskRetrievedFromPolicy.isEmpty()) {
-                return numberOfTaskStarted;
+            return getNumberOfTaskStarted(currentPolicy, jobMap, freeResources, fullListOfTaskRetrievedFromPolicy);
+        } finally {
+            if (toUnlock != null) {
+                schedulingService.unlockJobsToSchedule(toUnlock.values());
             }
+        }
+    }
+
+    private int getNumberOfTaskStarted(Policy currentPolicy, Map<JobId, JobDescriptor> jobMap,
```

Comment:
![BLOCKER](https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER') Refactor this method to not always return the same value. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3516)